enzyme/Enzyme/include/enzyme/declarations.hpp

```diff
@@ -0,0 +1,12 @@
+#pragma once
+
+extern int enzyme_dup;
+extern int enzyme_dupnoneed;
+extern int enzyme_out;
+extern int enzyme_const;
+
+template <typename return_type, typename... T>
+return_type __enzyme_fwddiff(void *, T...);
```

Comment:
We could also template the function type.
```suggestion
template <typename return_type, typename F, typename... T>
return_type __enzyme_fwddiff(F, T...);
```

Reply:
Out of curiosity, is it feasible to make `__enzyme_fwddiff` (and similar) support stateful lambdas directly? At present, users can differentiate stateful lambdas through a wrapper: https://fwd.gymni.ch/cKDhZ5 but it places more effort on the user to adapt their code to meet Enzyme's requirements.

Reply:
ok lets make it a pointer then.
```suggestion
template <typename return_type, typename F, typename... T>
return_type __enzyme_fwddiff(F*, T...);
```

Reply:
I think so, but it would end up looking more similar to the Julia API for Enzyme. Specifically, if there is state, one may need to have a const vs duplicated vs active (aka enzyme_out) version of the lambda.

Reply:
in practice, I think stateful lambdas would need to be implicitly `enzyme_const` w.r.t. the lambda member variables, since (in C++) there's no natural way to copy a lambda and access its member variables.

Reply:
Yeah the julia default is to have them const, unless explicitly passed a duplicated one. Unfortunately, it may be necessary to have a dup version if used as a temporary, or one otherwise wants to diff wrt.